### ISSUE TITLE
SIGSEGV when using `spawn_with_handle()`

### ISSUE THREAD

**Issue by realcr** [rust-lang-nursery/futures-rs#1471]
Hi, I am developing a proxy for a spawner that allows to intercept any spawned future and detect when no more progress can happen, a small project called [wait_spawner](https://github.com/freedomlayer/wait_spawner).

I bumped into an issue with `spawn_with_handle()` in one of my tests:

```rust
$ RUST_BACKTRACE=1 cargo test test_spawn_with_handle
    Finished dev [unoptimized + debuginfo] target(s) in 0.05s
     Running target/debug/deps/wait_spawner-6768db4e80bdfc39

running 1 test
error: process didn't exit successfully: `/home/real/projects/d/wait_spawner/target/debug/deps/wait_spawner-6768db4e80bdfc39 test_spawn_with_handle` (signal: 11, SIGSEGV: invalid memory reference)
```

I have no unsafe code, so I suspect that is some soundness problem that is related to `spawn_with_handle`, though I'm not sure.

## Versions
```
$ rustc --version
rustc 1.35.0-nightly (f22dca0a1 2019-03-05)
```
In my Cargo.toml:
```
futures-preview = {version = "0.3.0-alpha.13", features = ["compat"] }
```

## Reproducing the problem

The code could be found here:
https://github.com/freedomlayer/wait_spawner/tree/real/feat/spawn_with_handle_issue

To reproduce the issue, run the following:
```bash
$ git clone -b real/feat/spawn_with_handle_issue https://github.com/freedomlayer/wait_spawner
$ cargo test test_spawn_with_handle
```

## Other speculations

In some earlier version of my code I could detect a different kind of problem with `spawn_with_handle`.
It seems like `spawn_with_handle` will somehow attempt to `wake()` a future even after it has resolved. I was sure that `spawn_with_handle()` is just a wrapper over `spawn()`, but later I found that it really behaves differently.

**Comment by Nemo157** [comment]
I get an assertion failure when running that test (macOS 10.14, `rustc 1.34.0-nightly (f66e4697a 2019-02-20)`):

```
thread 'wait_spawner::tests::test_spawn_with_handle' panicked at 'assertion failed: !self.done.contains(&id)', src/wait_spawner.rs:76:9
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:39
   1: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:70
   2: std::panicking::default_hook::{{closure}}
             at src/libstd/sys_common/backtrace.rs:58
             at src/libstd/panicking.rs:200
   3: std::panicking::default_hook
             at src/libstd/panicking.rs:209
   4: <std::panicking::begin_panic::PanicPayload<A> as core::panic::BoxMeUp>::get
             at src/libstd/panicking.rs:478
   5: alloc::alloc::dealloc
             at /rustc/f66e4697ae286985ddefc53c3a047614568458bb/src/libstd/panicking.rs:412
   6: wait_spawner::wait_spawner::Tracker::insert
             at src/wait_spawner.rs:76
   7: <wait_spawner::wait_spawner::ArcWakerWrapper as futures_util::task::arc_wake::ArcWake>::wake
             at src/wait_spawner.rs:290
   8: futures_util::task::arc_wake::wake_arc_raw
             at /Users/nemo157/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-preview-0.3.0-alpha.13/src/task/arc_wake.rs:61
   9: <futures_channel::oneshot::Inner<T>>::drop_rx
             at /Users/nemo157/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-channel-preview-0.3.0-alpha.13/src/oneshot.rs:318
  10: <futures_channel::oneshot::Receiver<T> as core::ops::drop::Drop>::drop
             at /Users/nemo157/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-channel-preview-0.3.0-alpha.13/src/oneshot.rs:427
  11: core::ptr::real_drop_in_place
             at /rustc/f66e4697ae286985ddefc53c3a047614568458bb/src/libcore/ptr.rs:193
  12: core::ptr::real_drop_in_place
             at /rustc/f66e4697ae286985ddefc53c3a047614568458bb/src/libcore/ptr.rs:193
  13: futures_executor::local_pool::LocalPool::run_until
             at /Users/nemo157/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-executor-preview-0.3.0-alpha.13/src/local_pool.rs:158
  14: futures_executor::thread_pool::ThreadPool::run
             at /Users/nemo157/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-executor-preview-0.3.0-alpha.13/src/thread_pool.rs:98
  15: wait_spawner::wait_spawner::tests::test_spawn_with_handle
             at src/wait_spawner.rs:572
  16: wait_spawner::wait_spawner::tests::test_spawn_with_handle::{{closure}}
             at src/wait_spawner.rs:563
  17: core::ops::function::FnOnce::call_once
             at /rustc/f66e4697ae286985ddefc53c3a047614568458bb/src/libcore/ops/function.rs:231
  18: <F as alloc::boxed::FnBox<A>>::call_box
             at src/libtest/lib.rs:1497
             at /rustc/f66e4697ae286985ddefc53c3a047614568458bb/src/libcore/ops/function.rs:231
             at /rustc/f66e4697ae286985ddefc53c3a047614568458bb/src/liballoc/boxed.rs:739
  19: panic_unwind::dwarf::eh::read_encoded_pointer
             at src/libpanic_unwind/lib.rs:87
  20: test::run_test::run_test_inner::{{closure}}
             at /rustc/f66e4697ae286985ddefc53c3a047614568458bb/src/libstd/panicking.rs:276
             at /rustc/f66e4697ae286985ddefc53c3a047614568458bb/src/libstd/panic.rs:388
             at src/libtest/lib.rs:1452
```

I'll look a little more into the actual behaviour of your `wait_spawner` to see what's happening.

**Comment by Nemo157** [comment]
I think the issue that causes that assertion is the following order of events:

1. `spawn_with_handle` spawns a new task for the `ready` future
2. The new task runs
    a. The task checks if the receiver was dropped, registering its waker
    b. The task polls the `ready` future, it completes
    c. The task puts the result of the `ready` in the sender
    d. The task is complete
3. `thread_pool.run(handle)` runs
    a. The remote handle checks the receiver and gets the finished value
    b. The task drops the handle now that it's complete
    c. While dropping the receiver marks itself as dropped and wakes the sender in case it is blocked

You don't correctly handle this last wake after the future it is waking was completed, it's entirely valid to keep a waker and attempt to wake a future that has already been dropped.

**Comment by cramertj** [comment]
Closing, as this was addressed in #1491
